building-epics/cross-compile-epics-and-a-ioc-to-an-old-x86-linux.md

```diff
@@ -0,0 +1,367 @@
+
+sudo apt-get install gawk bison flex texinfo automake libtool cvs libcurses5-dev build-essential
+
+### Compiling
+
+In order to compile:
+
+```
+cd crosstool-ng-1.9.3
+./configure --local
+make
+```
+
+### Configuring
+
+Before configuring Crosstool-NG, I gathered information about my target system:
+
+Kernel Version:
+
+```
+$uname -r
+2.6.27.27
+```
+
+GCC Version:
+
+```
+$gcc --version
+gcc (GCC) 4.2.4
+```
+
+glibc version:
+
+```
+$/lib/libc.so.6
+GNU C Library stable release version 2.7, by Roland McGrath et al.
+```
+
+binutils version:
+
+```
+$ld --version
+GNU ld (Linux/GNU Binutils) 2.18.50.0.9.20080822
+```
+
+Based on this information, and on a lot of trial and error, I configured Crosstool-NG as follows (everything else set as default):
+
+```
+./ct-ng menuconfig
+
+PATHS AND MISC OPTIONS
+   \[\*\] Use obsolete features
+   
+TARGET OPTIONS
+   Target architecture (x86)
+   Bitness: (32-bit)
+   (i686) Architecture Level
+   
+OPERATING SYSTEM
+   Target OS (linux)
+   Linux kernel version (2.6.27.55)
+   
+BINARY UTILITIES
+   Binutils version (2.17)
+   
+C-COMPILER
+   GCC version (4.2.4)
+   \[\*\] C++
+   
+C-LIBRARY
+   glibc version (2.6.1)
+   
+```
+
+I tried other configurations, but they crashed the compilation process.
+
+### Compiling the toolchain
+
+Now we can compile the toolchain:
+
+```
+./ct-ng build
+```
+
+This will take a while. Go get some coffee or watch a cat video on Youtube.
+
+Once built, the toolchain will be in $HOME/x-tools/i686-unknown-linux-gnu/
+
+It’s a good idea (I think) to put the cross-compiler binaries on your path. Add this to the end of your ~/.bashrc:
+
+```
+PATH=$PATH:$HOME/x-tools/i686-unknown-linux-gnu/bin
+```
+
+Then source your .bashrc, so the changes take effect.
+
+```
+. ~/.bashrc
+```
+
+## EPICS dependencies
+
+In order to properly build epics-base to our target system, we have to take care of EPICS dependencies first. Namely, the libraries ‘readline’ and ‘ncurses’.
+
+They will be installed in our toolchain directory. We have to make it writable:
+
+```
+chmod -R +w $HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr
+```
+
+Now we can download, configure, compile and install the libraries.
+
+### readline
+
+```
+wget ftp://ftp.cwru.edu/pub/bash/readline-6.2.tar.gz
+tar -xvf readline-6.2.tar.gz
+cd readline-6.2
+./configure --prefix=$HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr --host=i686-unknown-linux-gnu
+make
+make install
+```
+
+### ncurses
+
+```
+wget ftp://ftp.gnu.org/pub/gnu/ncurses/ncurses-5.9.tar.gz
+tar -xvf ncurses-5.9.tar.gz
+cd ncurses-5.9
+./configure --prefix=$HOME/x-tools/i686-unknown-linux-gnu/i686-unknown-linux-gnu/sys-root/usr --host=i686-unknown-linux-gnu
+make
+make install
+```
+
+## Configure cross-compilation in EPICS
+
+We’re almost done. Back to the epics-base directory, open the file:
+
+```
+$EPICS\_BASE/configure/CONFIG\_SITE
```

Comment:
Here is an example of incorrect escapes: `EPICS_BASE` vs `EPICS\_BASE`. Could also be on the same line...